internal/backend/sessions/sessionworkflows/workflows.go

```diff
@@ -195,32 +193,42 @@ func (ws *workflows) sessionWorkflow(wctx workflow.Context, params *sessionWorkf
 	//                blow us up due to non determinism.
 
 	var prints []string
+	var duration time.Duration
+	var startTime time.Time
+	fields := []zap.Field{}
 
 	data, err := ws.getSessionData(wctx, params.SessionID)
 	if err == nil {
-		if workflow.IsReplaying(wctx) && data.Session.State().IsFinal() {
-			// HACK: If we somehow get a signal to this workflow after it completed,
-			//       in certain delicate timing, the workflow rekicks in replay.
-			//       Though replay would work just fine, and result in a no-op,
-			//       it's nice to not have to run through its entirely. Temporal
-			//       just seems to ignore it...
-			z.Info("already completed")
-			return nil, nil
-		}
 
-		sessionsCreatedCounter.Add(ctx, 1)
+		if workflow.IsReplaying(wctx) {
+			if data.Session.State().IsFinal() {
+				// HACK: If we somehow get a signal to this workflow after it completed,
+				//       in certain delicate timing, the workflow rekicks in replay.
+				//       Though replay would work just fine, and result in a no-op,
+				//       it's nice to not have to run through its entirely. Temporal
+				//       just seems to ignore it...
+				l.Info("already completed")
+				return nil, nil
+			}
+			fields = append(fields, zap.Int32("attempt", wi.Attempt))
+		} else {
+			sessionsCreatedCounter.Add(ctx, 1)
+		}
 
-		z.Info("session workflow: started")
-		prints, err = runWorkflow(wctx, z, ws, data, params.Debug)
+		startTime = data.Session.CreatedAt()
+		l.Info("session workflow: started", fields...)
+		prints, err = runWorkflow(wctx, l, ws, data, params.Debug)
+		duration = time.Since(startTime)
+		sessionDurationHistogram.Record(ctx, duration.Milliseconds())
 	}
 
 	if err != nil {
-		z := z.With(zap.Error(err))
+		sl := l.With(zap.Error(err))
```

Comment:
you can just add the duration here since you are creating a new logger anyway, instead of adding the duration all over below
or even before the error so it works for the success cases as well

Reply:
fixed